components/TextFields/src/ContainedInputView/private/MDCTextControlStyleFilled.m

```diff
@@ -0,0 +1,435 @@
+
+@property(strong, nonatomic) CAShapeLayer *filledSublayer;
+@property(strong, nonatomic) CAShapeLayer *thinUnderlineLayer;
+@property(strong, nonatomic) CAShapeLayer *thickUnderlineLayer;
+
+@property(strong, nonatomic, readonly, class) NSString *thickUnderlineGrowKey;
+@property(strong, nonatomic, readonly, class) NSString *thickUnderlineShrinkKey;
+@property(strong, nonatomic, readonly, class) NSString *thinUnderlineGrowKey;
+@property(strong, nonatomic, readonly, class) NSString *thinUnderlineShrinkKey;
+
+@property(strong, nonatomic) NSMutableDictionary<NSNumber *, UIColor *> *underlineColors;
+@property(strong, nonatomic) NSMutableDictionary<NSNumber *, UIColor *> *filledBackgroundColors;
+
+@end
+
+@implementation MDCTextControlStyleFilled
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    [self commonMDCTextControlStyleFilledInit];
+  }
+  return self;
+}
+
+- (void)commonMDCTextControlStyleFilledInit {
+  [self setUpUnderlineColors];
+  [self setUpFilledBackgroundColors];
+  [self setUpSublayers];
+}
+
+- (void)setUpUnderlineColors {
+  self.underlineColors = [NSMutableDictionary new];
+  UIColor *underlineColor = [UIColor blackColor];
+  self.underlineColors[@(MDCTextControlStateNormal)] = underlineColor;
+  self.underlineColors[@(MDCTextControlStateEditing)] = underlineColor;
+  self.underlineColors[@(MDCTextControlStateDisabled)] = underlineColor;
+}
+
+- (void)setUpFilledBackgroundColors {
+  self.filledBackgroundColors = [NSMutableDictionary new];
+  UIColor *filledBackgroundColor = [[UIColor blackColor] colorWithAlphaComponent:(CGFloat)0.1];
+  self.filledBackgroundColors[@(MDCTextControlStateNormal)] = filledBackgroundColor;
+  self.filledBackgroundColors[@(MDCTextControlStateEditing)] = filledBackgroundColor;
+  self.filledBackgroundColors[@(MDCTextControlStateDisabled)] = filledBackgroundColor;
+}
+
+- (void)setUpSublayers {
+  self.filledSublayer = [[CAShapeLayer alloc] init];
+  self.filledSublayer.lineWidth = 0.0;
+  self.thinUnderlineLayer = [[CAShapeLayer alloc] init];
+  [self.filledSublayer addSublayer:self.thinUnderlineLayer];
+  self.thickUnderlineLayer = [[CAShapeLayer alloc] init];
+  [self.filledSublayer addSublayer:self.thickUnderlineLayer];
+}
+
+#pragma mark Accessors
+
+- (UIColor *)underlineColorForState:(MDCTextControlState)state {
+  return self.underlineColors[@(state)];
+}
+
+- (void)setUnderlineColor:(nonnull UIColor *)underlineColor forState:(MDCTextControlState)state {
+  self.underlineColors[@(state)] = underlineColor;
+}
+
+- (UIColor *)filledBackgroundColorForState:(MDCTextControlState)state {
+  return self.filledBackgroundColors[@(state)];
+}
+
+- (void)setFilledBackgroundColor:(nonnull UIColor *)filledBackgroundColor
+                        forState:(MDCTextControlState)state {
+  self.filledBackgroundColors[@(state)] = filledBackgroundColor;
+}
+
+- (void)applyStyleToTextControl:(id<MDCTextControl>)textControl {
+  if (![textControl isKindOfClass:[UIView class]]) {
```

Comment:
Is this required by all of our textControls?
perhaps we should enforce that at the API / compiler level rather than at run time.

Proposed change to the protocol to
`- (void)applyStyleToTextControl:(UIView<MDCTextControl> *)textControl;`

Reply:
Good idea 👍